core/app/models/concerns/spree/user_address_book.rb

```diff
@@ -10,22 +10,22 @@ def find_first_by_address_values(address_attrs)
           detect { |ua| ua.address == Spree::Address.new(address_attrs) }
         end
 
-        # @note this method enforces only one default address per user
-        def mark_default(user_address)
+        def mark_default(user_address, address_kind = :shipping)
           # the checks of persisted? allow us to build a User and associate Addresses at once
+          default_flag = address_kind == :shipping ? :default : :default_billing
           ActiveRecord::Base.transaction do
             (self - [user_address]).each do |ua| # update_all would be nice, but it bypasses ActiveRecord callbacks
-              ua.persisted? ? ua.update!(default: false) : ua.default = false
+              ua.persisted? ? ua.update!({ default_flag => false }) : ua.write_attribute(default_flag, false)
             end
-            user_address.persisted? ? user_address.update!(default: true, archived: false) : user_address.default = true
+            user_address.persisted? ? user_address.update!({ default_flag => true, :archived => false }) : user_address.write_attribute(default_flag, true)
           end
         end
       end
 
       has_many :addresses, through: :user_addresses
 
-      # bill_address is only minimally used now, but we can't get rid of it without a major version release
-      belongs_to :bill_address, class_name: 'Spree::Address', optional: true
+      has_one :default_user_bill_address, ->{ default_billing }, class_name: 'Spree::UserAddress', foreign_key: 'user_id'
+      has_one :bill_address, through: :default_user_bill_address, source: :address
 
       has_one :default_user_address, ->{ default }, class_name: 'Spree::UserAddress', foreign_key: 'user_id'
       has_one :default_address, through: :default_user_address, source: :address
```

Comment:
I redefined them so to be respect symmetric with the other two. I kept `default_address=` defined as it is because the present behaviour calls `#save_in_address_book` with `default = true`, unlike `:ship_address=` and `bill_address=`.  `default_address_attributes=` calls `default_address=` so it stays too.